### PR TITLE
Database Streaming

### PR DESCRIPTION
**Pull request by ricdezen**
**fix (`postgresql_adapter.py`):**
**~** Was renamed from `posgresql_adapter`, both the module and the class got fixed. The reference in `upload.py` got fixed too.

**feat (`postgresql_adapter.py`):**
**+** Added `stream(query)` method to return a `PostgreSQLStream` object

**feat (`database_stream.py`):**
**+** **DatabaseStream**, generic Iterable class for database streams: `__next__()`, `__iter__()`, `has_next()`
**+** **PostgreSQLStream**: postgresql implementation:
**+** Requires a named cursor and the query to execute.
**+** Starts the query on the first `__iter__()` call and returns itself
**+** Each `__next__()` call calls `cursor.fetchone()`, when `StopIteration` is raised, closes the cursor and raises it again
**+** `has_next()` returns the opposite of `cursor.closed`
**+** `close()`, allows closing the cursor.

### CODE REVIEW

File: otri/database/database_stream.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
+import psycopg2
+
+from .database_query import DatabaseQuery
+
+
+class DatabaseStream():
+
+    def __iter__(self):
```
Comment (CremaLuca) [inline otri/database/database_stream.py]: Is this class an `Iterator` or an `Iterable`? You shouldn't mix the two things by using both `__iter__` and `__next__`

File: otri/database/postgresql_adapter.py [hunk 2]
```diff
@@ -1,11 +1,13 @@
 from .database_adapter import DatabaseAdapter, DatabaseData, DatabaseQuery
+from .database_stream import PostgreSQLStream
 
 import json
 import psycopg2
 from psycopg2.extras import execute_values
+from typing import Iterable
 
 
-class PosgreSQLAdapter(DatabaseAdapter):
```
Comment (CremaLuca) [inline otri/database/postgresql_adapter.py:8]: Nice, I'm a freaking dyslexic

File: otri/database/database_stream.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        Except:
+            StopIteration
+                When no result is given by the database (the query is over),
+                closes the cursor and raises StopIteration again.
+        '''
+        try:
+            return next(self.__cursor)
+        except StopIteration:
+            self.__cursor.close()
+            raise StopIteration
```
Comment (CremaLuca) [inline otri/database/database_stream.py:59]: You handled the cursor thing excellently, couldn't have done better

File: otri/database/postgresql_adapter.py [hunk 4]
```diff
@@ -72,9 +73,25 @@ def read(self, query: DatabaseQuery):
...
+    def stream(self, query: DatabaseQuery, batch_size: int = 1000) -> PostgreSQLStream:
+        '''
+        Returns a database stream.
+
+        Parameters:
+            query : DatabaseQuery
+                Executes the query on the given query.category and the given query.filters
+            batch_size : int
+                The number of rows the database should load before making them available.
+                The iterable still always yealds a single item.
```
Comment (CremaLuca) [inline otri/database/postgresql_adapter.py]:
```suggestion
                The iterable still always yields a single item.
```

File: run_all_tests.py [hunk 5]
```diff
@@ -1,3 +1,3 @@
 import os
 
-os.system('cmd /c "python -m pytest"')
\ No newline at end of file
+os.system('cmd /c "python -m pytest -s"')
```
Comment (CremaLuca) [inline run_all_tests.py:3]: What does this -s do?

File: otri/database/database_stream.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    def has_next(self):
```
Comment (CremaLuca) [inline otri/database/database_stream.py]:
```suggestion
    def has_next(self) -> bool:
```

File: otri/database/database_stream.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                A named cursor that can stream the given query.
+            query : DatabaseQuery
+                The query to stream.
+        '''
+        super().__init__()
+        self.__conn = connection
+        self.__query = query
+        self.__batch_size = batch_size
+
+    def __iter__(self):
```
Comment (CremaLuca) [inline otri/database/database_stream.py:94]:
```suggestion
    def __iter__(self) -> DatabaseIterator:
```

**Review by CremaLuca — APPROVED** [review]
Very noice, seems almost ready to be merged

File: run_all_tests.py (hunk 5 above)
Comment (ricdezen) [reply run_all_tests.py:3]: Removes the capturing of the print() statements, to allow logging

**Comment by CremaLuca** [comment]
Now it's-a perfect.
Perfect

**Pull request merged**